### ISSUE TITLE
ADR public key for DCR

### ISSUE THREAD

**Issue by jakubvozarik** [cdr-register/register#92]
The JWKS URI exposed by ADR gives the ADR’s public key which a data holder can use to validate the signature of the DCR(create register) Request. Subsequently, during modify/get/delete DCR requests, a data holder needs to authenticate the ADR using the ADR’s public key to validate the private_key_jwt. How does a data holder obtain the ADR’s public key during these operations. Should data holder store the ADR’s public key which it got it during ‘create register’ request?
a.            If so, how does the data holder make sure any update to ADR’s public key is propagated to all data holders? Because data holders have cached the public key initially during “create register” request.
b.            If not, should data holder use the “kid” parameter in private_key_jwt which comes in the request for “modify/get/delete” register operations and hit the JWKS URI again? This is an overhead because this call will happen for all token endpoint calls.

**Comment by perlboy** [comment]
> The JWKS URI exposed by ADR gives the ADR’s public key which a data holder can use to validate the signature of the DCR(create register) Request. 

This sentence is contradictory. The JWKS URI exposes a persistent URI to obtain an up to date JWKS. No transformation or permanence of the public key should or could be done because doing so would preclude the ability for an ADR to do key rotation. Essentially the JWKS URI is used to obtain the appropriate public keys for bearer token issuance. Bearer token issuance is governed by standard OAuth2 `POST /token` methods, in essence DCR components are resources available via this token.

>Subsequently, during modify/get/delete DCR requests, a data holder needs to authenticate the ADR using the ADR’s public key to validate the private_key_jwt. How does a data holder obtain the ADR’s public key during these operations. Should data holder store the ADR’s public key which it got it during ‘create register’ request?

Like with any other `private_key_jwt` call the Holder would retrieve it from the clients known `jwks_uri` location. The mechanism for this is the same regardless of if this is a DCR action or a consent initiation (ie. posting a Request Object) but the specification does outline a desire (or maybe a requirement, can't remember) to use different keys for the processes (ie. list multiple public keys in the JWKS).

> a. If so, how does the data holder make sure any update to ADR’s public key is propagated to all data holders? Because data holders have cached the public key initially during “create register” request.

A data holder doesn't care about other data holders because the client registration is unique per Holder/Recipient pair. The data holder retrieves the key set from the JWKS every request. 

> b. If not, should data holder use the “kid” parameter in private_key_jwt which comes in the request for “modify/get/delete” register operations and hit the JWKS URI again? This is an overhead because this call will happen for all token endpoint calls.

That's right it will and it does. This is normal OIDC and is why the JWKS is explicitly separated within discovery documents so that a CDN can be deployed for this relatively static content. The JWKS can be optionally cached for a "period of time", the underlying specifications aren't explicit about this but industry accepted norms are <1 hr. So for an ADR to be "sure" of propagation they publish a new key and don't start using it for an hour.

I believe the ACCC have designated 6 hrs as a maximum cache period for registration data so presumably a recipient might consider that an upper bound. 

In summary:

1. A holder should store the jwks_uri not the derived public key during registration as the public key should be allowed to be rotated (via different `kid`'s)
2. A holder is expected to use a recent copy of the JWKS retrievable from the registration URI
3. A holder is allowed to cache it for a reasonable period of time. It isn't explicitly defined by the specs but <6 hrs seems like an absolute upper bound ACCC/CDS may wish to consider declaring if it's a problem
